cartographer/io/internal/mapping_state_serialization.cc

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/io/internal/mapping_state_serialization.h"
+#include "cartographer/mapping/proto/serialization.pb.h"
+#include "cartographer/transform/transform.h"
+
+namespace cartographer {
+namespace io {
+
+namespace {
+
+mapping::proto::AllTrajectoryBuilderOptions
+CreateAllTrajectoryBuilderOptionsProto(
+    const std::vector<mapping::proto::TrajectoryBuilderOptionsWithSensorIds>&
+        all_options_with_sensor_ids) {
+  mapping::proto::AllTrajectoryBuilderOptions all_options_proto;
+  for (const auto& options_with_sensor_ids : all_options_with_sensor_ids) {
+    *all_options_proto.add_options_with_sensor_ids() = options_with_sensor_ids;
+  }
+  CHECK_EQ(all_options_with_sensor_ids.size(),
+           all_options_proto.options_with_sensor_ids_size());
+  return all_options_proto;
+}
+
+mapping::proto::SerializationHeader CreateHeader() {
+  mapping::proto::SerializationHeader header;
+  header.set_format_version(kMappingStateSerializationFormatVersion);
+  return header;
+}
+
+}  // namespace
+
+void ToPbStream(
+    const mapping::PoseGraph& pose_graph,
+    const std::vector<mapping::proto::TrajectoryBuilderOptionsWithSensorIds>&
+        trajectory_builder_options,
+    ProtoStreamWriterInterface* const writer) {
+  writer->WriteProto(CreateHeader());
+
+  mapping::proto::SerializedData proto;
+  // We serialize the pose graph followed by all the data referenced in it.
+  *proto.mutable_pose_graph() = pose_graph.ToProto();
+  writer->WriteProto(proto);
+  *proto.mutable_all_trajectory_builder_options() =
+      CreateAllTrajectoryBuilderOptionsProto(trajectory_builder_options);
+  writer->WriteProto(proto);
+
+  // Next serialize all submaps.
+  for (const auto& submap_id_data : pose_graph.GetAllSubmapData()) {
+    auto* const submap_proto = proto.mutable_submap();
+    submap_proto->mutable_submap_id()->set_trajectory_id(
+        submap_id_data.id.trajectory_id);
+    submap_proto->mutable_submap_id()->set_submap_index(
+        submap_id_data.id.submap_index);
+    submap_id_data.data.submap->ToProto(submap_proto,
+                                        /*include_probability_grid_data=*/true);
+    writer->WriteProto(proto);
```

Comment:
I would wrap these blocks into functions, that way there won't be a need for comments and also for {} blocks.

Reply:
done